### ISSUE TITLE
CDN route service (orig)

### ISSUE THREAD

**Issue by mogul** [18F/cg-agent-q#1]
What we're after: Applications can take advantage of platform-brokered CDN&nbsp;services

Benefits/why:
- Reduce costs for serving static content and assets
- Reduce ATO burden for tenants

Implementation notes:
- Generally, we're talking about AWS CloudFront, but avoid having tenants need to know much about that in case we have to change how we provide it in future; present the service to them generically, and avoid them having to learn about CloudFront in order to use it as much as possible.
- Leverage the [Cloud Foundry Service Broker API](http://docs.cloudfoundry.org/services/api.html)&nbsp;to expose provisioning of the service, independent of what UI/CLI is actually used by users... Ask for the necessary parameters in JSON up front, use [asynchronous operations](http://docs.cloudfoundry.org/services/api.html#asynchronous-operations)&nbsp;to talk about the state of the provisioning if it's going to take a while or is hung up on something external, etc.
- Here’s the [starting point for route services](http://docs.cloudfoundry.org/services/route-services.html), here’s [a reference on how to use `cf register-service-broker` to register ​"space-scoped private brokers"​](http://docs.cloudfoundry.org/services/managing-service-brokers.html), which you can use to test brokers as you work on them, and here's&nbsp; [a friendlier introduction](https://blog.starkandwayne.com/2015/11/18/register-your-own-service-broker-with-any-cloud-foundry/)&nbsp;to this idea.

**Comment by mogul** [comment]
Before we discussed and said we'd like this to be a route service, Diego was proposing it as a standalone thing to be done to develop SRE skills. He'd sketched this:

> - Login with UAA
> - CRUD CloudFront distributions with a desired url to use as origin (add a `/.well-known/` behavior that points to S3 for Lets encrypt validation)
> - Present instructions for a CNAME to be created to the right CloudFront distribution and wait for confirmation that the CNAME was created
> - Using the Lets Encrypt API request a TLS certificate and upload the validation to the S3 bucket for well-known files
> - Once the certificate is approved, upload it to IAM and set it so the CloudFront distribution serves it forwarding HTTP traffic to HTTPS.

Diego was nervous about this interaction being something that could possibly be encapsulated in a (relatively non-interactive) route service, due to the CNAME "handshake" in the middle; users don't know enough detail to point the CNAME at the right CloudFront distribution at the beginning of the process when they're providing parameters. 

I'm suggesting that the asynchronous `state: in progress` status could be paired with (for example) `description: waiting for CNAME to be created; see https://docs.cloud.gov/services/cdn#setting-up-cname`. Then the cloud.gov deck can surface this message to users whenever their service state is `in progress`.

Note also on this part:

> (add a `/.well-known/` behavior that points to S3 for Lets encrypt validation)

...there's a sketch for mapping the `/.well-known` subdirectory on existing routes to an app handling that aspect to be gleaned from the [`cf-letsencrypt`](https://github.com/bsyk/cf-letsencrypt) repository; may be useful.

**Comment by afeld** [comment]
@jmcarp and I talked about this today, and took some notes here:

https://docs.google.com/document/d/1NknEm7aTLXujuGZnWuWjLouVCLo3DrJXMFYjaqnjKLs/edit

**Comment by afeld** [comment]
...so sounds like we have a plan! Per some [conversation in Slack](https://18f.slack.com/archives/compliance-toolkit/p1462996242001051), sounds like I'm going to working on Masonry things for a week or two, so might be mostly @jmcarp running with this for now.

**Comment by mogul** [comment]
@jmcarp What else is needed to call this done and announce it as a feature? Is it documented for users/operators?

**Comment by jmcarp** [comment]
@mogul: a few projects have used this successfully in production, so I think we're all right to enable access across all organizations and add it to the docs. That said, there are a few tech debt issues that I would like to address at some point so that we can feel a little more confident in the service.

**Comment by mogul** [comment]
OK, going to consider it "in-progress but not done".

**Comment by mogul** [comment]
For reasons not worth going into, I've had to create a separate issue in GitHub for this one linked to Aha!... It's [here](https://github.com/18F/cg-product/issues/86). I already reparented the sub-issues onto that one. Please subscribe to that one and I'll close this one.
